cdap-storage-spi/src/main/java/co/cask/cdap/spi/data/transaction/TransactionRunners.java

```diff
@@ -216,6 +242,30 @@ private static RuntimeException propagate(Throwable throwable) {
     throw propagate(cause);
   }
 
+  /**
+   * Propagates the given {@link TransactionException}. If the {@link TransactionException#getCause()}
+   * doesn't return {@code null}, the cause will be used instead for the propagation. This method will
+   * throw the failure exception as-is the given propagated types if the type matches or as {@link RuntimeException}.
+   *
+   * @param e the {@link TransactionException} to propagate
+   * @param propagateType1 if the exception is an instance of this type, it will be rethrown as is
+   * @param propagateType2 if the exception is an instance of this type, it will be rethrown as is
+   * @param propagateType3 if the exception is an instance of this type, it will be rethrown as is
+   * @param <X1> exception type of first propagate type
+   * @param <X2> exception type of second propagate type
+   * @param <X3> exception type of second propagate type
+   */
+  public static <X1 extends Throwable, X2 extends Throwable, X3 extends Throwable> X1
+  propagate(TransactionException e,
+            Class<X1> propagateType1,
+            Class<X2> propagateType2,
+            Class<X3> propagateType3) throws X1, X2, X3 {
```

Comment:
Is there a better way to propagate exceptions? Seems like we will have to add similar methods if more than 3 exception types need to be thrown.

Reply:
Since the throws clause is static, I don't see an easy way around adding new methods. We can think about a different way when we feel the list is unwieldily and confusing. I don't think we are at that point right now.